pkg/operator/ceph/nfs/nfs_test.go

```diff
@@ -0,0 +1,209 @@
+		assert.NoError(t, err)
+
+		assert.NotEqual(t, hash1, hash2)
+	})
+
+	t.Run("running with different configs should give different hashes", func(t *testing.T) {
+		cmName, hash1, err := r.createConfigMap(nfs, "a")
+		assert.NoError(t, err)
+		assert.Equal(t, "rook-ceph-nfs-my-nfs-a", cmName)
+		_, err = r.context.Clientset.CoreV1().ConfigMaps("rook-ceph-test-ns").Get(context.TODO(), cmName, metav1.GetOptions{})
+		assert.NoError(t, err)
+
+		nfs2 := nfs.DeepCopy()
+		nfs2.Name = "nfs-two"
+		_, hash2, err := r.createConfigMap(nfs2, "a")
+		assert.NoError(t, err)
+		_, err = r.context.Clientset.CoreV1().ConfigMaps("rook-ceph-test-ns").Get(context.TODO(), cmName, metav1.GetOptions{})
+		assert.NoError(t, err)
+
+		assert.NotEqual(t, hash1, hash2)
+	})
+}
+
+func TestReconcileCephNFS_upCephNFS(t *testing.T) {
+	ns := "up-ceph-ns-namespace"
+
+	s := scheme.Scheme
+
+	clientset := k8sfake.NewSimpleClientset()
+
+	executor := &exectest.MockExecutor{
+		MockExecuteCommandWithOutput: func(command string, args ...string) (string, error) {
+			logger.Infof("executing command: %s %+v", command, args)
+			if args[0] == "auth" {
+				if args[1] == "get-or-create-key" {
+					return "{\"key\":\"mysecurekey\"}", nil
+				}
+			}
+			panic(errors.Errorf("unhandled command %s %v", command, args))
+		},
+	}
+
+	client := fake.NewClientBuilder().WithScheme(s).WithRuntimeObjects().Build()
+
+	c := &clusterd.Context{
+		Executor:  executor,
+		Client:    client,
+		Clientset: clientset,
+	}
+
+	r := &ReconcileCephNFS{
+		scheme:  s,
+		context: c,
+		clusterInfo: &cephclient.ClusterInfo{
+			FSID:        "myfsid",
+			CephVersion: cephver.Octopus,
+			Context:     context.TODO(),
+			Namespace:   ns,
+		},
+		cephClusterSpec: &cephv1.ClusterSpec{
+			CephVersion: cephv1.CephVersionSpec{
+				Image: "quay.io/ceph/ceph:v15",
+			},
+		},
+	}
+
+	nfs := &cephv1.CephNFS{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-nfs",
+			Namespace: ns,
+		},
+		Spec: cephv1.NFSGaneshaSpec{
+			RADOS: cephv1.GaneshaRADOSSpec{
+				Pool:      "myfs-data0",
+				Namespace: "nfs-test-ns",
+			},
+			Server: cephv1.GaneshaServerSpec{
+				Active: 2,
+			},
+		},
+	}
+
+	err := r.upCephNFS(nfs)
+	assert.NoError(t, err)
+
+	deps, err := r.context.Clientset.AppsV1().Deployments(ns).List(context.TODO(), metav1.ListOptions{})
+	assert.NoError(t, err)
+	assert.Len(t, deps.Items, 2)
+	names := []string{}
+	hashes := []string{}
+	for _, dep := range deps.Items {
+		names = append(names, dep.Name)
+		assert.Contains(t, dep.Spec.Template.Annotations, "config-hash")
+		hashes = append(hashes, dep.Spec.Template.Annotations["config-hash"])
+	}
+	assert.ElementsMatch(t, []string{"rook-ceph-nfs-my-nfs-a", "rook-ceph-nfs-my-nfs-b"}, names)
+	assert.NotEqual(t, hashes[0], hashes[1])
+
+	svcs, err := r.context.Clientset.CoreV1().Services(ns).List(context.TODO(), metav1.ListOptions{})
+	assert.NoError(t, err)
+	// Each NFS server gets a service.
```

Comment:
This is the current behavior, but I'm not sure this is right. Should this be the design?

Reply:
This is part of the design doc: 

> The operator creates a k8s service for each of the ganesha server pods
  to allow each of the them to have a stable IP address.